experimental/gradually-typed/src/Torch/GraduallyTyped/Examples/TwoLayerNetwork.hs

```diff
@@ -0,0 +1,407 @@
+  _ =>
+  -- | device
+  SDevice device ->
+  -- | model specification
+  ModelSpec model ->
+  -- | model state dict
+  StateDict ->
+  -- | input
+  [[Float]] ->
+  -- | expected output
+  [[Float]] ->
+  -- | random generator
+  Generator generatorDevice ->
+  -- | loss
+  m (_, Generator generatorOutputDevice)
+computeLoss device modelSpec stateDict input expectedOutput =
+  runIxStateT $
+    ilift
+      ( let inputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device input
+            expectedOutputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device expectedOutput
+            model = flip evalStateT stateDict $ fromStateDict modelSpec mempty
+         in (,,) <$> inputTensor <*> expectedOutputTensor <*> model
+      )
+      >>>= ( \(inputTensor, expectedOutputTensor, model) ->
+               ireturn inputTensor
+                 >>>= IxStateT . forward model
+                 >>>= ilift . (`mseLoss` expectedOutputTensor)
+           )
+
+train ::
+  _ =>
+  -- | random seed for the training
+  Word64 ->
+  -- | device
+  SDevice device ->
+  -- | model specification
+  ModelSpec model ->
+  -- | model state dict
+  StateDict ->
+  -- | learning rate
+  Double ->
+  -- | stream of training examples
+  P.ListT IO ([[Float]], [[Float]], int) ->
+  IO StateDict
+train seed device modelSpec stateDict learningRate examples = do
+  g <- sMkGenerator device seed
+  optim <- mkAdam defaultAdamOptions {learningRate} (Map.elems stateDict)
+  let step g' (x, y, _iter) = do
+        let f :: ForeignPtr ATen.TensorList -> IO (ForeignPtr ATen.Tensor)
+            f tensorList = do
+              stateDict' :: StateDict <- ATen.uncast tensorList (pure . Map.fromList . zip (Map.keys stateDict))
+              UnsafeTensor t <- fst <$> computeLoss device modelSpec stateDict' x y g'
+              pure t
+        _loss :: ForeignPtr ATen.Tensor <- ATen.cast2 ATen.step optim f
+        print $ UnsafeTensor _loss
+        pure g'
+      init' = pure g
+      done = pure
+  _ <- P.foldM step init' done . P.enumerate $ examples
+  pure stateDict
+
+-- | Single-cycle learning rate schedule.
+-- See, for instance, https://arxiv.org/abs/1803.09820.
+singleCycleLearningRateSchedule :: Double -> Double -> Int -> Int -> Int -> Int -> Double
+singleCycleLearningRateSchedule maxLearningRate finalLearningRate numEpochs numWarmupEpochs numCooldownEpochs epoch
+  | epoch <= 0 = 0.0
+  | 0 < epoch && epoch <= numWarmupEpochs =
+    let a :: Double = fromIntegral epoch / fromIntegral numWarmupEpochs
+     in a * maxLearningRate
+  | numWarmupEpochs < epoch && epoch < numEpochs - numCooldownEpochs =
+    let a :: Double =
+          fromIntegral (numEpochs - numCooldownEpochs - epoch)
+            / fromIntegral (numEpochs - numCooldownEpochs - numWarmupEpochs)
+     in a * maxLearningRate + (1 - a) * finalLearningRate
+  | otherwise = finalLearningRate
+
+runTwoLayerNetworkExample :: IO ()
+runTwoLayerNetworkExample = do
```

Comment:
I think this is the main function of two layer network, and runTwoLayerNetworkExample is not called by test.
Should we use ghci to call it?

Reply:
Perhaps. It would need to be optimized for speed, this runs quite slowly currently.

It's also not clear where this code should live long-term. Perhaps I'll move it to experimental/gradually-typed/examples/two-layer-network and make an executable out of it.

I've added this code mostly to get a feel for how training will look like, and what abstractions will be useful.

There will be more examples. @jul1u5 wants to do one on MNIST, and I want to do one with transformers, likely BERT and T5.